src/gs-discovery-feed-content-provider.c

```diff
@@ -0,0 +1,296 @@
+		return cache_filename;
+	}
+
+    return cache_filename;
+}
+
+static const guint N_APPS_TO_INCLUDE_IN_RESULT = 5;
+
+static void
+search_done_cb (GObject *source,
+		GAsyncResult *res,
+		gpointer user_data)
+{
+	PendingSearch *search = user_data;
+	GsDiscoveryFeedContentProvider *self = search->provider;
+	guint i;
+	GVariantBuilder builder;
+	guint app_list_length;
+	guint count;
+	g_autoptr(GsAppList) list = NULL;
+
+	list = gs_plugin_loader_search_finish (self->plugin_loader, res, NULL);
+	if (list == NULL) {
+	    printf("No applications returned, cannot continue");
+		g_dbus_method_invocation_return_value (search->invocation, g_variant_new ("(as)", NULL));
+		pending_search_free (search);
+		g_application_release (g_application_get_default ());
+		return;
+	}
+
+	app_list_length = gs_app_list_length (list);
+	count = 0;
+
+	/* sort by kudos, as there is no ratings data by default */
+	gs_app_list_sort (list, search_sort_by_kudo_cb, NULL);
+
+	g_variant_builder_init (&builder, G_VARIANT_TYPE ("aa{sv}"));
+	for (i = 0; i < app_list_length; i++) {
+	    gchar *app_thumbnail_filename;
+	    GdkPixbuf *pixbuf;
+	    GVariant *icon_serialized = NULL;
+		GsApp *app = gs_app_list_index (list, i);
+
+		if (gs_app_get_state (app) != AS_APP_STATE_AVAILABLE)
+			continue;
+
+		/* This app must have an icon which is serializable */
+		pixbuf = gs_app_get_pixbuf (app);
+		if (pixbuf != NULL)
+			icon_serialized = g_icon_serialize (G_ICON (pixbuf));
+
+		if (!icon_serialized)
+			continue;
+
+		/* Check to make sure that the app has an available thumbnail
+		 * otherwise, we need to skip it completely */
+		app_thumbnail_filename = get_app_thumbnail_cached_filename (app);
+
+		if (!app_thumbnail_filename)
+			continue;
+
+		g_variant_builder_open (&builder, G_VARIANT_TYPE("a{sv}"));
+		g_variant_builder_add (&builder, "{sv}", "app_id", g_variant_new_string (gs_app_get_id (app)));
+		g_variant_builder_add (&builder, "{sv}", "id", g_variant_new_string (gs_app_get_id (app)));
+		g_variant_builder_add (&builder, "{sv}", "name", g_variant_new_string (gs_app_get_name (app)));
+		g_variant_builder_add (&builder, "{sv}", "synopsis", g_variant_new_string (gs_app_get_summary (app)));
+		g_variant_builder_add (&builder, "{sv}", "thumbnail_uri", g_variant_new_string (app_thumbnail_filename));
+		g_variant_builder_add (&builder, "{sv}", "icon", icon_serialized);
+		g_free (app_thumbnail_filename);
+		g_variant_unref (icon_serialized);
+
+		g_variant_builder_close (&builder);
+
+		if (++count > N_APPS_TO_INCLUDE_IN_RESULT)
+			break;
+	}
+	g_dbus_method_invocation_return_value (search->invocation,
+	                                       g_variant_new ("(aa{sv})", &builder));
+
+	pending_search_free (search);
+	g_application_release (g_application_get_default ());
+}
+
+static gboolean
+handle_get_discovery_feed_apps (GsDiscoveryFeedInstallableApps 	   *skeleton,
```

Comment:
The alignment for the arguments looks a bit inconsistent here. There are a few other places as well.